stock_ownership_availability_rules/model/quant.py

```diff
@@ -0,0 +1,59 @@
+from openerp import models, api, fields
+
+
+class Quant(models.Model):
+    _inherit = 'stock.quant'
+
+    @api.model
+    def create(self, vals):
+        """Set the owner based on the location.
+
+        This is not a method because we need to know the location.
+
+        """
+        if not vals.get('owner_id'):
+            Company = self.env['res.company']
+            location = self.env['stock.location'].browse(vals['location_id'])
+
+            vals['owner_id'] = (
+                location.partner_id.id
+                or location.company_id.partner_id.id
+                or Company.browse(
+                    Company._company_default_get('stock.quant')
+                ).partner_id.id
+            )
+
+        return super(Quant, self).create(vals)
+
+    owner_id = fields.Many2one('res.partner', 'Owner',
+                               help="This is the owner of the quant",
+                               readonly=True,
+                               select=True,
+                               required=True)
+
+    @api.model
+    def quants_get_prefered_domain(self, location, product, qty, domain=None,
+                                   prefered_domain_list=None,
+                                   restrict_lot_id=False,
+                                   restrict_partner_id=False):
+        if domain is None:
+            domain = []
+        if prefered_domain_list is None:
+            prefered_domain_list = []
+
+        my_partner = location.company_id.partner_id
+        if restrict_partner_id == my_partner.id or not restrict_partner_id:
```

Comment:
Several points here:

1. i find the identifier 'my_partner' confusing in this context. 
2. nothing seems to be done about location.partner_id.id? Is it normal?
3. the condition in the test is not clear to me. I would have expected `and restrict_partner_id` instead of `or not restrict_partner_id`. Am I missing something obvious?